presto-spi/src/main/java/com/facebook/presto/spi/block/BlockFlattener.java

```diff
@@ -49,6 +54,16 @@ public BlockLease flatten(Block block)
         return newLease(block);
     }
 
+    public void reset()
```

Comment:
This method is dangerous. If the flattener was created with a shared allocator it will release all arrays that might be still in use.

I think a better solution is to change OptimizedPartitionedOutputOperator#decodeBlock to take an additional parameter, Closer, and add `blockLease::close` to it. Then, call this closer in the place where you currently call reset.

Reply:
@mbasmanova thanks for your suggestion. Could you be more specific? Did you mean we store a blockLease in every DecodedBlockNode, and call blockLease.close() recursively on every decodedBlock at the end of each page?

Reply:
@yingsu00 

I suggest to keep track of all the leases produced by the flattener and close then when no longer needed. Here is how this might look like:

```
        private final BlockFlattener flattener = new BlockFlattener();
        private final Closer blockLeaseCloser = Closer.create();
...
            for (int i = 0; i < decodedBlocks.length; i++) {
                decodedBlocks[i] = decodeBlock(flattener, blockLeaseCloser, page.getBlock(i));
            }
...
    static DecodedBlockNode decodeBlock(BlockFlattener flattener, Closer blockLeaseCloser, Block block)
    {
        BlockLease lease = flattener.flatten(block);
        blockLeaseCloser.register(lease::close);
        Block decodedBlock = lease.get();

        if (decodedBlock instanceof ArrayBlock) {
            ColumnarArray columnarArray = ColumnarArray.toColumnarArray(decodedBlock);
            return new DecodedBlockNode(columnarArray, ImmutableList.of(decodeBlock(flattener, blockLeaseCloser, columnarArray.getElementsBlock())));
        }

        if (decodedBlock instanceof MapBlock) {
            ColumnarMap columnarMap = ColumnarMap.toColumnarMap(decodedBlock);
            return new DecodedBlockNode(columnarMap, ImmutableList.of(decodeBlock(flattener, blockLeaseCloser, columnarMap.getKeysBlock()), decodeBlock(flattener, blockLeaseCloser, columnarMap.getValuesBlock())));
        }

        if (decodedBlock instanceof RowBlock) {
            ColumnarRow columnarRow = ColumnarRow.toColumnarRow(decodedBlock);
            ImmutableList.Builder<DecodedBlockNode> children = ImmutableList.builder();
            for (int i = 0; i < columnarRow.getFieldCount(); i++) {
                children.add(decodeBlock(flattener, blockLeaseCloser, columnarRow.getField(i)));
            }
            return new DecodedBlockNode(columnarRow, children.build());
        }

        if (decodedBlock instanceof DictionaryBlock) {
            return new DecodedBlockNode(decodedBlock, ImmutableList.of(decodeBlock(flattener, blockLeaseCloser, ((DictionaryBlock) decodedBlock).getDictionary())));
        }

        if (decodedBlock instanceof RunLengthEncodedBlock) {
            return new DecodedBlockNode(decodedBlock, ImmutableList.of(decodeBlock(flattener, blockLeaseCloser, ((RunLengthEncodedBlock) decodedBlock).getValue())));
        }

        return new DecodedBlockNode(decodedBlock, ImmutableList.of());
    }
...
            // Close block leases
            try {
                blockLeaseCloser.close();
            }
            catch (IOException e) {
                throw new UncheckedIOException(e);
            }

Reply:
@mbasmanova Thank you very much for the suggestion! I updated the commit 32ef717604 Fix "invalid position in block" error in the repartition PR https://github.com/prestodb/presto/pull/13183. Now this PR just have one commit "Add getRetainedSizeInBytes to BlockFlattener". Do you think this one looks good and can be merged?